File: kotlin-node/karakum/src/jsMain/resources/nameResolvers/resolveBufferConstantsName.js
import ts from "typescript";
import * as karakum from "../karakum.mjs";

export default (node, context) => {
    const typeScriptService = context.lookupService(karakum.typeScriptServiceKey.get())
    const getParent = typeScriptService?.getParent.bind(typeScriptService) ?? (node => node.parent)

    const sourceFileName = node.getSourceFile()?.fileName ?? "generated.d.ts"
    if (!sourceFileName.endsWith("buffer.d.ts")) return null

    const variable = getParent(node)
    if (!variable) return null
    if (!ts.isVariableDeclaration(variable)) return null
    if (!ts.isIdentifier(variable.name)) return null

    const variableName = variable.name.text
    if (variableName !== "constants") return null

    return `BufferConstants`
}

File: kotlin-node/karakum/src/jsMain/kotlin/node/karakum/nameResolvers/resolveBufferConstantsName.kt
package node.karakum.plugins

import io.github.sgrishchenko.karakum.extension.Context
import io.github.sgrishchenko.karakum.extension.plugins.TypeScriptService
import io.github.sgrishchenko.karakum.extension.plugins.typeScriptServiceKey
import io.github.sgrishchenko.karakum.util.getSourceFileOrNull
import node.karakum.util.nullable
import typescript.Node
import typescript.isIdentifier
import typescript.isVariableDeclaration

fun resolveBufferConstantsName(node: Node, context: Context) = nullable {
    val typeScriptService = ensureNotNull(context.lookupService<TypeScriptService>(typeScriptServiceKey))

    val sourceFileName = ensureNotNull(node.getSourceFileOrNull()).fileName
    ensure(sourceFileName.endsWith("buffer.d.ts"))

    val variable = ensureNotNull(typeScriptService.getParent(node))
    ensure(isVariableDeclaration(variable))

    val variableName = variable.name
    ensure(isIdentifier(variableName))
    ensure(variableName.text == "constants")

    "BufferConstants"
}
